lib/serialization/README.md

```diff
@@ -0,0 +1,237 @@
+# Abstract serialization services
+
+The serialization services are centered around the `auto_serializable` annotation,
+the `Serializable` interface and the implementations of `Serializer` and `Deserializer`.
+
+## The `auto_serializable` annotation
+
+A class annotated with `auto_serializable` identifies it as a subclass of Serializable and
+triggers the generation of customized serialization and deserialization services.
+
+~~~
+import serialization
+
+# Simple serializable class identifying a human
+class Person
+	auto_serializable
+
+	# First and last name
+	var name: String
+
+	# Year of birth (`null` if unknown)
+	var birth: nullable Int
+
+	redef fun ==(o) do return o isa SELF and name == o.name and birth == o.birth
+	redef fun hash do return name.hash
+end
+~~~
+
+The `Person` class also defines `==` and `hash`, this is optional but we will use it to make an important point.
+By definition of a serializable class, an instance can be serialized to a stream, then deserialized.
+The deserialized instance will not be the same instance, but they should be equal.
+So, in this case, we can compare both instances with `==` to test their equality.
+
+Some conditions applies to the classes that can be annotated as `auto_serializable`.
+All attributes of the class must be serializable, runtime errors will be
+raised when trying to serialize non-serializable attributes.
+
+In the class `Person`, all attributes are typed with classes the standards library.
+These common types are defined defined as serializable by this project.
+The attributes could also be typed with user-defined `auto_serializable`
+classes or any other subclass of `Serializable`.
+
+~~~
+# This `auto_serializable` class is composed of two `auto_serializable` attributes
+class Couple
+	auto_serializable
+
+	var partner_a: Person
+	var partner_b: Person
+
+	redef fun ==(o) do return o isa SELF and partner_a == o.partner_a and partner_b == o.partner_b
+	redef fun hash do return partner_a.hash + 1024*partner_b.hash
+end
+~~~
+
+The `auto_serializable` applies only to the class definition,
+only attributes declared locally will be serialized.
+However, each definition of a class (a refinement or specialization)
+can declare `auto_serializable`.
```

Comment:
It is correct, it is the current behavior. It respects the visibility of the setter of `B::j=`, and in a different module we may not see `A::i=`.

This specific case may break at runtime... it could probably be detected at compilation too. The precise error depends on the constructors. I believe that, on deserialization, this would not set `A::i`, so it will crash on access.

In practice, a class must define `auto_serializable` in each definition with attributes. This will force us to declare classes as `auto_serializable` in their introduction in the standard library.